src/Utilities/GridFileReader.f90

```diff
@@ -0,0 +1,521 @@
+    call hash_table_da(this%typ)
+    call hash_table_da(this%shp_idx)
+    deallocate (this%shp)
+  end subroutine finalize
+
+  !> @brief Read the file's self-describing header. Internal use only.
+  subroutine read_header(this)
+    class(GridFileReaderType) :: this
+    call this%read_header_meta()
+    call this%read_header_body()
+  end subroutine read_header
+
+  !> @brief Read self-describing metadata (first four lines). Internal use only.
+  subroutine read_header_meta(this)
+    ! dummy
+    class(GridFileReaderType) :: this
+    ! local
+    character(len=50) :: line
+    integer(I4B) :: lloc, istart, istop
+    integer(I4B) :: ival
+    real(DP) :: rval
+
+    ! grid type
+    read (this%inunit) line
+    lloc = 1
+    call urword(line, lloc, istart, istop, 1, ival, rval, 0, 0)
+    if (line(istart:istop) /= 'GRID') then
+      call store_error('Binary grid file must begin with "GRID". '//&
+                       &'Found: '//line(istart:istop))
+      call store_error_unit(this%inunit)
+    end if
+    call urword(line, lloc, istart, istop, 1, ival, rval, 0, 0)
+    this%grid_type = line(istart:istop)
+    call upcase(this%grid_type)
+
+    ! version
+    read (this%inunit) line
+    lloc = 1
+    call urword(line, lloc, istart, istop, 0, ival, rval, 0, 0)
+    call urword(line, lloc, istart, istop, 2, ival, rval, 0, 0)
+    this%version = ival
+
+    ! ntxt
+    read (this%inunit) line
+    lloc = 1
+    call urword(line, lloc, istart, istop, 0, ival, rval, 0, 0)
+    call urword(line, lloc, istart, istop, 2, ival, rval, 0, 0)
+    this%ntxt = ival
+
+    ! lentxt
+    read (this%inunit) line
+    lloc = 1
+    call urword(line, lloc, istart, istop, 0, ival, rval, 0, 0)
+    call urword(line, lloc, istart, istop, 2, ival, rval, 0, 0)
+    this%lentxt = ival
+
+  end subroutine read_header_meta
+
+  !> @brief Read the header body section (text following the first
+  !< four "meta" lines). Internal use only.
+  subroutine read_header_body(this)
+    ! dummy
+    class(GridFileReaderType) :: this
+    ! local
+    character(len=:), allocatable :: body
+    character(len=:), allocatable :: line
+    character(len=10) :: key, dtype
+    real(DP) :: rval
+    integer(I4B) :: i, lloc, istart, istop, ival
+    integer(I4B) :: nvars, ndim, dim, ishp
+    integer(I4B), allocatable :: shp(:)
+
+    allocate (this%keys(this%ntxt))
+    allocate (character(len=this%lentxt*this%ntxt) :: body)
+    allocate (character(len=this%lentxt) :: line)
+
+    nvars = 0
+    read (this%inunit) body
+    do i = 1, this%lentxt * this%ntxt, this%lentxt
+      line = body(i:i + this%lentxt - 1)
+      lloc = 1
+
+      ! key
+      lloc = 1
+      call urword(line, lloc, istart, istop, 1, ival, rval, 0, 0)
+      key = line(istart:istop)
+      call upcase(key)
+      nvars = nvars + 1
+      this%keys(nvars) = key
+
+      ! type
+      call urword(line, lloc, istart, istop, 1, ival, rval, 0, 0)
+      dtype = line(istart:istop)
+      call upcase(dtype)
```

Comment:
fyi -- urword has an argument (ncode) that will convert to upper case, so the call here to upcase is not necessary.